src/python/pants/engine/README.md

```diff
@@ -49,44 +50,44 @@ then the `@rule` will run exactly once, and the value that it produces will be a
 
 #### Example
 
-As a very simple example, you might register the following `@rule` that can compute a `String`
-Product given a single `Int` argument.
+As a very simple example, you might register the following `@rule` that can compute a `str`
+`Product` given a single `int` argument.
 
 ```python
-@rule(str, [int])
-def int_to_str(an_int):
-  return str(an_int)
+@rule
+def int_to_str(value: int) -> str:
+  return str(value)
 ```
 
-The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
-second argument is a list of "parameter selectors" that declare the types of the input parameters for
-the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
-selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
+Here the only requirements for the `@rule` are:
+1. The decorated function is a free, top-level function; ie: methods cannot be `@rule`s.
+2. The decorated function must be fully type annotated.
+3. As mentioned above, the types of the function `Param`s must be hashable and have useful equality.
```

Comment:
It's very subtle, but: a `@rule`s arguments are not necessarily exactly the `Params` it requires. See the second paragraph of the "Products and Params" section. In a few other places this document refers to "argument selectors", but we don't have a more formal distinction than that.

Reply:
Ok - I'll strike. In general I was trying to remove mention of engine internals, but this is a mized document already. Much of its language is imo noisy for rule writers, but important for engine hackers. I'll probably be circling in soon to break up the document along those lines, or otherwise extract the relevant bits and expand so @rule writers can get to business and not get confused unless they need to dive deep.

Reply:
That sounds like a really good idea! I've been linking to the engine README to others to try to explain "what it is", and I think the potential split you've described would allow us to have a document that could serve that purpose, while also keeping around the notes for engine hackers elsewhere.